### PR TITLE
Implement system fonts system channel listener

### PR DESCRIPTION
**Pull request by chunhtai**
## Description

This pr add a system channel listener to repaint RenderParagraph and RenderEditable upon receiving system fonts change notification. We will have engine side change coming soon

https://github.com/flutter/engine/pull/12276

## Related Issues

https://github.com/flutter/flutter/issues/38556

## Tests

I added the following tests:

RenderParagraph relayout upon system fonts changes
RenderEditable relayout upon system fonts changes

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I signed the [CLA].
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] All existing and new tests are passing.
- [ ] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [ ] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [ ] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Review by goderbauer — COMMENTED** [review]
What about other objects/widgets that use TextPainter to measure text (example: CupertinoDataPicker or Material's TimePicker)? Do we not have to notify them as well so they can rebuild?

File: packages/flutter/lib/src/rendering/binding.dart [hunk 1]
```diff
@@ -415,3 +415,20 @@ class RenderingFlutterBinding extends BindingBase with GestureBinding, ServicesB
     renderView.child = root;
   }
 }
+
+/// Client for listening to system fonts change.
+///
+/// Using this mixin will cause it automatically subscribe to system channel.
+mixin SystemFontsClient on RenderObject {
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/binding.dart]: Maybe call this "RelayoutWhenSystemFontsChangeMixin" to better describe what it does?

File: packages/flutter/lib/src/rendering/binding.dart [hunk 2]
```diff
@@ -415,3 +415,20 @@ class RenderingFlutterBinding extends BindingBase with GestureBinding, ServicesB
     renderView.child = root;
   }
 }
+
+/// Client for listening to system fonts change.
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/binding.dart]: "Mixin for [RenderObject] that will call [markNeedsLayout] whenever the system fonts change".

Can you also add an explanation why/when a developer should mix this in?

File: packages/flutter/lib/src/rendering/binding.dart [hunk 3]
```diff
@@ -415,3 +415,20 @@ class RenderingFlutterBinding extends BindingBase with GestureBinding, ServicesB
...
+///
+/// Using this mixin will cause it automatically subscribe to system channel.
+mixin SystemFontsClient on RenderObject {
+  @override
+  void attach(covariant Object owner) {
+    super.attach(owner);
+    PaintingBinding.instance?.systemFonts?.addListener(markNeedsLayout);
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/binding.dart]: Why would instance or systemFonts be null? Shouldn't they always be initialized when you get to this place?

File: packages/flutter/lib/src/rendering/binding.dart [hunk 4]
```diff
@@ -415,3 +415,20 @@ class RenderingFlutterBinding extends BindingBase with GestureBinding, ServicesB
...
+  }
+
+  @override
+  void detach() {
+    super.detach();
+    PaintingBinding.instance?.systemFonts?.removeListener(markNeedsLayout);
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/binding.dart]: Same here.

File: packages/flutter/lib/src/rendering/binding.dart (hunk 1 above)
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/binding.dart]: We should probably find a better location then `rendering/binding.dart` for this.

File: packages/flutter/lib/src/painting/binding.dart [hunk 5]
```diff
@@ -79,6 +81,42 @@ mixin PaintingBinding on BindingBase, ServicesBinding {
...
+  /// correctly reflect the change, it is important to relayout text related
+  /// widgets when this happens.
+  Listenable systemFonts;
+
+  Set<VoidCallback> _systemFontsCallbacks;
+
+  /// System channel message handler in painting layer.
+  ///
+  /// This handler checks for system fonts change notification and invokes
+  /// callbacks in [systemFontsOCallbacks].
```
Comment (goderbauer) [inline packages/flutter/lib/src/painting/binding.dart]: nit: spelling within [].

File: packages/flutter/lib/src/painting/binding.dart [hunk 6]
```diff
@@ -79,6 +81,42 @@ mixin PaintingBinding on BindingBase, ServicesBinding {
...
   }
+
+  /// Listenable for OSes system fonts change event.
+  ///
+  /// System fonts can change when system installs or removes new font. To
+  /// correctly reflect the change, it is important to relayout text related
+  /// widgets when this happens.
+  Listenable systemFonts;
+
+  Set<VoidCallback> _systemFontsCallbacks;
```
Comment (goderbauer) [inline packages/flutter/lib/src/painting/binding.dart]: Put this inside the _SystemFonts class?

File: packages/flutter/lib/src/painting/binding.dart [hunk 7]
```diff
@@ -79,6 +81,42 @@ mixin PaintingBinding on BindingBase, ServicesBinding {
...
+  ///
+  /// This handler checks for system fonts change notification and invokes
+  /// callbacks in [systemFontsOCallbacks].
+  @override
+  Future<void> handleSystemMessage(Object systemMessage) async {
+    await super.handleSystemMessage(systemMessage);
+    final Map<String, dynamic> message = systemMessage;
+    final String type = message['type'];
+    if (type == 'fontsChange') {
+      for (VoidCallback callback in _systemFontsCallbacks) {
```
Comment (goderbauer) [inline packages/flutter/lib/src/painting/binding.dart]: put this inside a "notifyListeners" method on _SystemFonts?

File: packages/flutter/lib/src/painting/binding.dart [hunk 8]
```diff
@@ -79,6 +81,42 @@ mixin PaintingBinding on BindingBase, ServicesBinding {
...
+        callback();
+      }
+    }
+  }
+}
+
+class _SystemFonts extends Listenable {
```
Comment (goderbauer) [inline packages/flutter/lib/src/painting/binding.dart]: Maybe _SystemFontsListener

File: packages/flutter/lib/src/painting/binding.dart (hunk 5 above)
Comment (goderbauer) [reply packages/flutter/lib/src/painting/binding.dart]: Actually, since this is an override you could just remove this doc comment altogether. It documents implementation details anyways.

File: packages/flutter/lib/src/cupertino/date_picker.dart [hunk 9]
```diff
@@ -100,6 +100,18 @@ class _DatePickerLayoutDelegate extends MultiChildLayoutDelegate {
   }
 }
 
+/// Retrieves cached widths of columns in the input state.
+@visibleForTesting
+Map<int, double> debugGetCupertinoDatePickerStateCache(State<CupertinoDatePicker> state) {
```
Comment (goderbauer) [inline packages/flutter/lib/src/cupertino/date_picker.dart]: This is awkward code to add to the framework. Can we find another way to test this?

File: packages/flutter/lib/src/cupertino/date_picker.dart [hunk 10]
```diff
@@ -447,6 +459,21 @@ class _CupertinoDatePickerDateTimeState extends State<CupertinoDatePicker> {
     }
 
     previousHourIndex = selectedHour;
+
+    PaintingBinding.instance.systemFonts.addListener(_handleSystemFontsChange);
+  }
+
+  void _handleSystemFontsChange () {
+    didChangeDependencies();
```
Comment (goderbauer) [inline packages/flutter/lib/src/cupertino/date_picker.dart]: calling didChangeDependecies seems awkward (especially because that one calls super). Maybe factor out what you need from didChangeDependencies into a seperate method that we can call fro both places?

File: packages/flutter/lib/src/cupertino/date_picker.dart [hunk 11]
```diff
@@ -774,6 +801,21 @@ class _CupertinoDatePickerDateState extends State<CupertinoDatePicker> {
     selectedYear = widget.initialDateTime.year;
 
     dayController = FixedExtentScrollController(initialItem: selectedDay - 1);
+
+    PaintingBinding.instance.systemFonts.addListener(_handleSystemFontsChange);
+  }
+
+  void _handleSystemFontsChange() {
+    didChangeDependencies();
```
Comment (goderbauer) [inline packages/flutter/lib/src/cupertino/date_picker.dart]: Same here.

File: packages/flutter/lib/src/services/binding.dart [hunk 12]
```diff
@@ -47,6 +49,11 @@ mixin ServicesBinding on BindingBase {
     return const _DefaultBinaryMessenger._();
   }
 
+  /// Handler for flutter/system message channel.
```
Comment (goderbauer) [inline packages/flutter/lib/src/services/binding.dart]: Handler called for messages received on the [SystemChannels.system] call.

Other bindings may override this to respond to incoming system message.

File: packages/flutter/lib/src/cupertino/date_picker.dart [hunk 13]
```diff
@@ -464,6 +464,21 @@ class _CupertinoDatePickerDateTimeState extends State<CupertinoDatePicker> {
     }
 
     previousHourIndex = selectedHour;
+
+    PaintingBinding.instance.systemFonts.addListener(_handleSystemFontsChange);
+  }
+
+  void _handleSystemFontsChange () {
+    estimatedColumnWidths.clear();
```
Comment (goderbauer) [inline packages/flutter/lib/src/cupertino/date_picker.dart]: Put this statement into the setState closure? (That's the state getting changed here, right?)

File: packages/flutter/lib/src/cupertino/date_picker.dart [hunk 14]
```diff
@@ -464,6 +464,21 @@ class _CupertinoDatePickerDateTimeState extends State<CupertinoDatePicker> {
...
+    setState(() {
+      // didChangeDependencies should update the width if necessary.
```
Comment (goderbauer) [inline packages/flutter/lib/src/cupertino/date_picker.dart]: This comment seems outdated?

File: packages/flutter/lib/src/cupertino/date_picker.dart [hunk 15]
```diff
@@ -791,6 +806,21 @@ class _CupertinoDatePickerDateState extends State<CupertinoDatePicker> {
     selectedYear = widget.initialDateTime.year;
 
     dayController = FixedExtentScrollController(initialItem: selectedDay - 1);
+
+    PaintingBinding.instance.systemFonts.addListener(_handleSystemFontsChange);
+  }
+
+  void _handleSystemFontsChange() {
+    _refreshEstimatedColumnWidths();
```
Comment (goderbauer) [inline packages/flutter/lib/src/cupertino/date_picker.dart]: Put this inside the setState since that's causing the state to change?

File: packages/flutter/lib/src/cupertino/date_picker.dart [hunk 16]
```diff
@@ -791,6 +806,21 @@ class _CupertinoDatePickerDateState extends State<CupertinoDatePicker> {
...
+    setState(() {
+      // didChangeDependencies should update the width if necessary.
```
Comment (goderbauer) [inline packages/flutter/lib/src/cupertino/date_picker.dart]: This comment seems outdated?

File: packages/flutter/lib/src/cupertino/date_picker.dart (hunk 13 above)
Comment (goderbauer) [inline packages/flutter/lib/src/cupertino/date_picker.dart]: Maybe add a comment why we have to clear this on system font changes (e.g. because it was determined by measuring the width of text which may have changed with the new fonts).

File: packages/flutter/lib/src/cupertino/date_picker.dart (hunk 15 above)
Comment (goderbauer) [inline packages/flutter/lib/src/cupertino/date_picker.dart]: Maybe add a comment here as well?

File: packages/flutter/lib/src/painting/binding.dart [hunk 17]
```diff
@@ -79,6 +81,45 @@ mixin PaintingBinding on BindingBase, ServicesBinding {
     super.evict(asset);
     imageCache.clear();
   }
+
+  /// Listenable for OSes system fonts change event.
+  ///
+  /// System fonts can change when system installs or removes new font. To
```
Comment (goderbauer) [inline packages/flutter/lib/src/painting/binding.dart]: when the system

File: packages/flutter/lib/src/painting/binding.dart [hunk 18]
```diff
@@ -79,6 +81,45 @@ mixin PaintingBinding on BindingBase, ServicesBinding {
...
+  /// correctly reflect the change, it is important to relayout text related
+  /// widgets when this happens.
```
Comment (goderbauer) [inline packages/flutter/lib/src/painting/binding.dart]: Maybe add a little more details here, e.g. objects that show text and/or measure text (e.g. via [TextPainter]) should listen to this and redraw/remeasure.

File: packages/flutter/lib/src/painting/binding.dart [hunk 19]
```diff
@@ -79,6 +81,45 @@ mixin PaintingBinding on BindingBase, ServicesBinding {
...
+  Listenable systemFonts;
```
Comment (goderbauer) [inline packages/flutter/lib/src/painting/binding.dart]: This shouldn't be writable for the public.

File: packages/flutter/lib/src/painting/binding.dart [hunk 20]
```diff
@@ -79,6 +81,45 @@ mixin PaintingBinding on BindingBase, ServicesBinding {
...
+
+  Set<VoidCallback> _systemFontsCallbacks;
```
Comment (goderbauer) [inline packages/flutter/lib/src/painting/binding.dart]: Can this be removed now?

File: packages/flutter/lib/src/rendering/editable.dart [hunk 21]
```diff
@@ -10,6 +10,7 @@ import 'package:flutter/gestures.dart';
 import 'package:flutter/semantics.dart';
 import 'package:flutter/services.dart';
 
+import 'binding.dart';
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/editable.dart]: This import can probably be removed?

File: packages/flutter/lib/src/rendering/object.dart [hunk 22]
```diff
@@ -3213,6 +3213,26 @@ mixin ContainerRenderObjectMixin<ChildType extends RenderObject, ParentDataType
   }
 }
 
+/// Mixin for [RenderObject] that will call [markNeedsLayout] whenever the
+/// system fonts change
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/object.dart]: add `.` at the end.

File: packages/flutter/lib/src/rendering/object.dart [hunk 23]
```diff
@@ -3213,6 +3213,26 @@ mixin ContainerRenderObjectMixin<ChildType extends RenderObject, ParentDataType
...
+///
+/// System fonts can change when OS install or remove a font. Use this mixin if
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/object.dart]: when the operating system installs or removes a font.

File: packages/flutter/lib/src/rendering/object.dart [hunk 24]
```diff
@@ -3213,6 +3213,26 @@ mixin ContainerRenderObjectMixin<ChildType extends RenderObject, ParentDataType
...
+/// the [RenderObject] uses [TextPainter] or text related painter to correctly
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/object.dart]: What's a text related painter?

File: packages/flutter/lib/src/rendering/paragraph.dart [hunk 25]
```diff
@@ -13,6 +13,7 @@ import 'package:flutter/services.dart';
 
 import 'package:vector_math/vector_math_64.dart';
 
+import 'binding.dart';
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/paragraph.dart]: Can this import be removed?

File: packages/flutter/lib/src/services/binding.dart [hunk 26]
```diff
@@ -47,6 +49,13 @@ mixin ServicesBinding on BindingBase {
     return const _DefaultBinaryMessenger._();
   }
 
+  /// Handler called for messages received on the [SystemChannels.system] call.
```
Comment (goderbauer) [inline packages/flutter/lib/src/services/binding.dart]: call -> message channel.

File: packages/flutter/lib/src/services/binding.dart [hunk 27]
```diff
@@ -47,6 +49,13 @@ mixin ServicesBinding on BindingBase {
...
+  ///
+  /// Other bindings may override this to respond to incoming system message.
```
Comment (goderbauer) [inline packages/flutter/lib/src/services/binding.dart]: message -> messages.

File: packages/flutter/lib/src/widgets/binding.dart [hunk 28]
```diff
@@ -556,15 +555,14 @@ mixin WidgetsBinding on BindingBase, SchedulerBinding, GestureBinding, RendererB
...
+  @override
+  Future<void> handleSystemMessage(Object systemMessage) async {
+    await super.handleSystemMessage(systemMessage);
     final Map<String, dynamic> message = systemMessage;
     final String type = message['type'];
-    switch (type) {
-      case 'memoryPressure':
-        handleMemoryPressure();
-        break;
+    if (type == 'memoryPressure') {
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/binding.dart]: Why the change from a switch?

File: packages/flutter/lib/src/painting/binding.dart (hunk 20 above)
Comment (chunhtai) [reply packages/flutter/lib/src/painting/binding.dart]: Yes, I forgot to remove it

File: packages/flutter/lib/src/rendering/object.dart (hunk 24 above)
Comment (chunhtai) [reply packages/flutter/lib/src/rendering/object.dart]: using ui.Paragraph directly without going through text painter. I should update the doc comment

**Comment by chunhtai** [comment]
@goderbauer  this is ready to review.

File: packages/flutter/lib/src/rendering/editable.dart [hunk 29]
```diff
@@ -632,6 +632,13 @@ class RenderEditable extends RenderBox {
     markNeedsLayout();
   }
 
+  @override
+  void systemFontsDidChange() {
+    super.systemFontsDidChange();
+    _textPainter.markNeedsLayout();
```
Comment (chunhtai) [inline packages/flutter/lib/src/rendering/editable.dart]: I am not sure how do we test _textPainter has been marked needs layout. this also apply to all the other widget.
The current test only verify renderobject is dirty after font change.

File: packages/flutter/lib/src/material/range_slider.dart [hunk 30]
```diff
@@ -930,6 +930,14 @@ class _RenderRangeSlider extends RenderBox {
     markNeedsLayout();
   }
 
+  @override
+  void systemFontsDidChange() {
+    super.systemFontsDidChange();
+    _startLabelPainter.markNeedsLayout();
```
[the remark this answers is not in the document]
Comment (chunhtai) [reply packages/flutter/lib/src/material/range_slider.dart]: no, _updateLabelPainters will set the same property of existing textpainter, which will all get shortcircuited.

File: packages/flutter/lib/src/rendering/object.dart [hunk 31]
```diff
@@ -3213,6 +3213,40 @@ mixin ContainerRenderObjectMixin<ChildType extends RenderObject, ParentDataType
...
+/// system fonts change.
+///
+/// System fonts can change when the OS install or remove a font. Use this mixin if
+/// the [RenderObject] uses [TextPainter] or [Paragraph] to correctly update the
+/// text when it happens.
+mixin RelayoutWhenSystemFontsChangeMixin on RenderObject {
+
+  /// A callback that is called when system fonts have changed.
+  ///
+  /// By default, [mark needsLayout] is called on the [RenderObject]
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/object.dart]: nit: no space after mark. (sorry...)

**Review by goderbauer — APPROVED** [review]
LGTM
